src/util/flatRender.js

```diff
@@ -0,0 +1,192 @@
+
+    const isHidden = children.some(child => child.value === '=hidden')
+    const isMeta = isFunction(value)
+
+    // hide if this node is itself a meta function or has children meta =hidden
+    // if showHiddenThoughts is true then don't hide at all
+    const shouldHide =
+      !showHiddenThoughts &&
+      (isMeta || isHidden)
+
+    const depthInfo = {
+      hiddenNodes: acc.depthInfo.hiddenNodes + (isHidden ? 1 : 0),
+      metaNodes: acc.depthInfo.metaNodes + (isMeta ? 1 : 0),
+    }
+
+    // do not recurse if hidden
+    if (!showDistantAncestor || shouldHide) return { ...acc, depthInfo }
+
+    // siblings count which are not meta functions
+    const siblingsCountWithoutMeta = subThoughts.reduce(
+      (acc, child) => acc + (!isFunction(child.value) ? 1 : 0),
+      0
+    )
+
+    // stop deeper recursion at certain depth where any descendant of cursor has more than one subthought
+    // stop further deeper recursion if max depth is reached
+    const stop =
+      addDistantAncestorAndStop ||
+      (isCursorDescendant && siblingsCountWithoutMeta > 1) ||
+      childPath.length - cursor.length === MAX_DEPTH_FROM_CURSOR
+
+    const distanceFromCursor = cursor.length - childPath.length
+
+    // if true the node will have reduced opacity on render
+    const isDistantThought =
+      (!isLeaf
+        ? distanceFromCursor >= 0
+        : distanceFromCursor >= (isCursorAncestor ? 2 : 1)) && !isCursor
+
+    const recursiveUpdate = getFlatArray({
+      startingPath: childPath,
+      children,
+      cursor,
+      isLeaf,
+      showHiddenThoughts,
+      isParentCursorAncestor: isCursorAncestor,
+      isCursorDescendant: isCursorDescendant || isCursor,
+    })
+
+    const flatArrayDescendants = stop
+      ? []
+      : recursiveUpdate.flatArray
+
+    const childrenDepthInfo = recursiveUpdate.depthInfo
+
+    /**
+     * This is the logic for showing '▸' or '•' i.e if there would be any visible nodes if we expand this node
+     *
+     * 1. If there are visible nodes returned from recursive call then show expand icons.
+     * 2. Else if children length is less than zero then show '•'.
+     * 3. Else if all children nodes are either meta nodes or a node that has hidden attribute (=hidden),
+     *    then only show expand icon '▸' if showHiddenThoughts is true else show '•'.
+     *
+     * For example
+     *
+     * When showHiddenThoughts is true
+     *
+     *    ▸ A
+     *      ▸ B
+     *        ▸ C
+     *          • D
+     *          • =hidden
+     *        • =immovable
+     *
+     * If showHiddenThoughts is false then thought B despite having two childrens won't render anthing.
+     *
+     *    ▸ A
+     *      ▸ B
+     *        • C (So instead of '▸' we show '•')
+     *
+     * */
+
+    const hasChildren = flatArrayDescendants.length > 0 || (
+      children.length > 0 &&
+      ((childrenDepthInfo.hiddenNodes + childrenDepthInfo.metaNodes) !== children.length ? true : showHiddenThoughts)
+    )
```

Comment:
Is there a case where `flatArrayDescendants.length > 0` is `true` and the rest of the logic is `false`? If not, you can omit `flatArrayDescendants.length > 0`.

I tested it with this and it never happened:

```js
    const a = flatArrayDescendants.length > 0
    const b = children.length > 0 &&
      ((childrenDepthInfo.hiddenNodes + childrenDepthInfo.metaNodes) !== children.length ? true : showHiddenThoughts)
    if (a && !b) {
      console.log(childPath) // never called?
    }
```

Reply:
@raineorshine  Yes it makes sense. We can omit `flatArrayDescendants.length > 0`